aoc/d3/main.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+// split a string in two using the middle point
+func SplitString(input string) []string {
+	half := len(input) / 2
+	return []string{input[:half], input[half:]}
+}
+
+// find the character that exists in two strings
+func FindCommonItem(input1 string, input2 string) string {
+	var result string
+	for _, item := range input1 { // generated by ChatGPT
+		if i := strings.Index(input2, string(item)); i != -1 {
+			result = string(item)
+		}
+	}
+	return result
+}
+
+// return the item priority
+// given the value scheme for letters, I think this is a good way of getting the value
+// rather than storing every letter in a struct
+func GetItemPriority(item string) int {
+	alphabet := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	value := strings.Index(alphabet, item)
+	return value + 1 // offset as a=1 and not 0
+}
+
+// calculate total duplicate item value in each rucksack
+func GetTotalItemValue(rucksacks []string) int {
+	total := 0
+	for _, rucksack := range rucksacks {
+		compartments := SplitString(rucksack)
+		common_item := FindCommonItem(compartments[0], compartments[1])
+		total = total + GetItemPriority(common_item)
+	}
+	return total
+}
+
+// calculate total duplicate item value for given file
+func p1() {
+	//file_path := "aoc/d3/test.txt"
+	file_path := "aoc/d3/submission.txt"
+	file_string, _ := os.ReadFile(file_path)
+	rucksacks := strings.Split(string(file_string), "\n")
+	total := GetTotalItemValue(rucksacks)
+	fmt.Println(total)
+}
+
+// not used yet
+func p2() {
+	//file_path := "aoc/d2/test.txt"
+	file_path := "aoc/d2/submission.txt"
+	os.ReadFile(file_path)
+}
+
+// main function to run the function given as argument
+func main() {
+	function_name := os.Args[1]
+	fmt.Println(function_name)
+	funcs := map[string]func(){
+		"p1": p1,
+		"p2": p2,
+	}
+	funcs[function_name]()
```

Comment:
ngl, I don't like this 🤣 I'd recommend using https://pkg.go.dev/flag and making `main.go` hold only this function. You can create another file with `package main` and they all see each other as if they were in the same file. Well you can't have multiple packages in a single directory anyway but you get my drift.

Also, probably more performant to just have:

```suggestion
	function_name := os.Args[1]
	fmt.Println(function_name)
	switch function_name {
	case "p1":
	    p1()
	case "p2":
	    p2()
	default:
	    log.Fatal("What are you doing fool?")
	}
```
Otherwise you're building a map, accessing it and then calling. The switch statement on it's own is probably faster than creating the map. Having said that, I haven't measured it so I could just be chatting shit

Reply:
Okay it was pretty dirty I admit. I found a really easy way to do this in python by using `locals()[func]`. I've changed it to a flag and using a switch case like you suggested. I would like the two possibilities to remain in the one file as that's how this advent of code is organized. It has one problem and then a second problem which makes you change/reuse your existing code. Changed in https://github.com/lucasrijllart/advent-of-code-2022/pull/1/commits/00e8af19a283703a1f9fe098974fd2303a1a8fa5